hyperspy/learn/mva.py

```diff
@@ -1317,22 +1369,679 @@ def _estimate_elbow_position(self, curve_values):
+
+                for i,k in enumerate(k_range):
+                    alg = self._cluster_analysis(k,
+                                                 scaled_data,
+                                                 algorithm,
+                                                 **kwargs)
+
+                    D = distances_within_cluster(scaled_data,alg.labels_)
+                    W = np.sum(D)
+                    inertia[i]= np.log(W)
+                    pbar.update(1)
+                    _logger.info("For n_clusters =", k,
+                                 "The distance metric is :",
+                                 inertia[-1])
+                    to_return = inertia
+                    best_k =self._estimate_elbow_position(to_return,log=False)\
+                        +min_k
+            elif metric == "silhouette":
+                k_range   = list(range(2, max_clusters+1))
+                pbar = progressbar(total=len(k_range))
+                silhouette_avg = []
+                for k in k_range:
+                    alg = self._cluster_analysis(k,
+                                                 scaled_data,
+                                                 algorithm,
+                                                 **kwargs)
+                    cluster_labels = alg.labels_
+                    silhouette_avg.append(
+                        import_sklearn.sklearn.metrics.silhouette_score(
+                        scaled_data,
+                        cluster_labels))
+                    pbar.update(1)
+                    _logger.info("For n_clusters =", k,
+                                 "The average silhouette_score is :",
+                                 silhouette_avg[-1])
+                to_return = silhouette_avg
+                best_k = []
+                max_value = -1.0
+                # find peaks
+                for u in range(len(silhouette_avg)-1):
+                    if ((silhouette_avg[u] > silhouette_avg[u-1]) &
+                            (silhouette_avg[u] > silhouette_avg[u+1])):
+                        best_k.append(u+min_k)
+                        max_value = max(silhouette_avg[u], max_value)
+                if silhouette_avg[0] > max_value:
+                    best_k.insert(0, min_k)
+            else:
+                # cluster and calculate gap statistic
+                # various empty arrays...
+                reference_inertia = np.zeros(len(k_range))
+                reference_std  = np.zeros(len(k_range))
+                data_inertia=np.zeros(len(k_range))
+                reference=np.zeros(scaled_data.shape)
+                local_inertia = np.zeros(n_ref)
+                pbar = progressbar(total=n_ref*len(k_range))
+                # only perform 1 pass of clustering
+                # otherwise std_dev isn't correct
+
+                for o_indx,k in enumerate(k_range):
+                    # calculate the data metric
+                    np.random.seed(1)
```

Comment:
Why can't we pass a `np.random.RandomState` object here? Or at the very least, be able to vary the seed.

Reply:
Removed np random seed - user can set it or pass random_state as a kwarg.